Update module loads to intel/2022 for jobs running gsi.x and enkf.x

RussTreadon-NOAA opened the pull request:
# Description
This PR changes the module loads in g-w jobs `anal.sh`, `eobs.sh`, and `eupd.sh` to use machine specific modulefiles from `sorc/gsi_enkf.fd/modulefiles`.  This is necessary because NOAA-EMC PR [#571](https://github.com/NOAA-EMC/GSI/pull/571) updated `gsi.x` and `enkf.x` builds to intel/2022 on NOAA RDHPCS machines.

Resolves #1863

**NOTE**:  This g-w PR depends GSI PR [#629](https://github.com/NOAA-EMC/GSI/pull/629).  Do not merge this g-w PR into `develop` until GSI PR [#629](https://github.com/NOAA-EMC/GSI/pull/629) is merged into NOAA-EMC/GSI `develop`.  This g-w PR needs to update the GSI hash in g-w `Externals.cfg` and `sorc/checkout.sh` 

# Type of change
- New feature (adds functionality)

# Change characteristics
The changes in this PR allow g-w to run `gsi.x` and `enkf.x` on NOAA RDHPCS machines using intel/2022.

# How has this been tested?
A cycled parallel covering the period 2021073106 through 2021080118 has been run on Hera, Orion, and WCOSS2 (Cactus).  All g-w successfully ran to completion.

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code

RussTreadon-NOAA committed:
add scripting to load gsi_enkf modules in anal.sh, eobs.sh, and eupd.sh (#1863)

RussTreadon-NOAA committed:
refactor module loads in anal.sh, eobs.sh, and eupd.sh to account for wcoss2 (#1863)

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/intel2022_updates

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/intel2022_updates

WalterKolczynski-NOAA commented on jobs/rocoto/anal.sh:
```diff
@@ -4,9 +4,36 @@ source "${HOMEgfs}/ush/preamble.sh"
+    # Append compiler (only on machines that have multiple compilers)
+    COMPILER=${COMPILER:-"intel"}
+    if [[ "${MACHINE_ID}" = "hera" ]] || [[ "${MACHINE_ID}" = "cheyenne" ]]; then
+	MACHINE_ID=${MACHINE_ID}.${COMPILER}
```
```suggestion
        MACHINE_ID=${MACHINE_ID}.${COMPILER}
```

Repeat for the other files to fix the indentation.

RussTreadon-NOAA replied:
Indention fixed at 2d4d020f.

WalterKolczynski-NOAA reviewed (changes requested):
I'm assuming this module mess is very temporary until we update the rest of workflow to Intel 2022 on RDHPCS machines (which I think was just waiting on GSI).

RussTreadon-NOAA commented:
> I'm assuming this is very temporary until we update the rest of workflow to Intel 2022 on RDHPCS machines (which I think was just waiting on GSI).

Yes, this is a temporary patch.  I asked about the status of an intel/2022 ready g-w.  @aerorahul provided an informative reply which motivated creation of issue #1863.

> ---------- Forwarded message ---------
From: Rahul Mahajan - NOAA Federal <rahul.mahajan@noaa.gov>
Date: Fri, Sep 15, 2023 at 11:44 AM
Subject: Re:
To: Russ Treadon - NOAA Federal <russ.treadon@noaa.gov>
Cc: Walter Kolczynski - NOAA Affiliate <Walter.Kolczynski@noaa.gov>, Kate Friedman - NOAA Federal <Kate.Friedman@noaa.gov>, Catherine Thomas - NOAA Federal <catherine.thomas@noaa.gov>, Arun Chawla - NOAA Federal <arun.chawla@noaa.gov>, Daryl Kleist - NOAA Federal <daryl.kleist@noaa.gov>



> Hi Russ,
> 
> Thank you for letting us know that the GSI has been updated to the newer Intel compiler.  This removes a big blocker in the work of the global-workflow towards the use of a consolidated and consistent software stack.  Something we have been waiting on for quite a while.
> 
> Dave Huber has already started working on the other DA related components GSI-utils, GSI-monitor to ensure they can be built and run with the same compiler as the GSI.  
> He is also ensuring that the other components of the GFS can use the same compiler as the GSI.
> As you can imagine, this is not something that can be accomplished in short order, but rest assured, work is underway and has been prioritized.  I cannot at this time give you an estimated date until Dave can perform a first assessment.  Kate also had done work to test with newer compilers in the global-workflow, but without the cycling, so we are hopeful this will also speed up the migration.
> 
> We had to short-circuit the ufs-weather-model as it had moved past the Intel 18 compiler some time ago.  If you look at the f[cst.sh](https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/rocoto/fcst.sh) and [efcs.sh](https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/rocoto/efcs.sh), we had to load the modules from the ufs-weather-model directly rather than sourcing load_fv3gfs_modules.sh.  If it is urgent for the DA development, I suggest taking that approach while Dave works on a holistic solution.
> 
> I hope this is satisfactory and allows the development to proceed.
> 
> Thanks,
> Rahul

WalterKolczynski-NOAA commented:
Sounds good. Just the indentation of that one line in each file and then the hash when it is ready.

RussTreadon-NOAA committed:
fix indentation innrocoto jobs (#1863)

DavidHuber-NOAA reviewed (approved):
Looks good, thanks @RussTreadon-NOAA!

WalterKolczynski-NOAA commented:
Converted to draft so it isn't accidentally merged before the hash gets updated.

RussTreadon-NOAA commented:
Thank you @WalterKolczynski-NOAA and @DavidHuber-NOAA .  I'll update the GSI hash one GSI PR [#629](https://github.com/NOAA-EMC/GSI/pull/629) is approved and merged into NOAA-EMC/GSI `develop`.   Good idea to change to draft.  I should have done so from the start.

RussTreadon-NOAA committed:
update GSI hash to bring in intel/2022 bug fixes and safeguards (#1863)

RussTreadon-NOAA commented:
GSI PR [#629](https://github.com/NOAA-EMC/GSI/pull/629) has been merged into NOAA-EMC/GSI `develop` at [`ca19008`](https://github.com/NOAA-EMC/GSI/commit/ca19008b521cccf07a84037cfece3847232025cc).  The GSI hash in `Externals.cfg` and `sorc/checkout.sh` have updated to [`ca19008`](https://github.com/NOAA-EMC/GSI/commit/ca19008b521cccf07a84037cfece3847232025cc).  

Done at [`a86b95a`](https://github.com/RussTreadon-NOAA/global-workflow/commit/a86b95ac6d9fa1c98dceb0838c718ed64fc1bfec).

This PR may be changed back to _ready for review_.

RussTreadon-NOAA commented:
The CI-Orion failure occurs when attempting to execute `getncdimlen`
```
++ JGLOBAL_ATMOS_ANALYSIS[52]: /work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/getncdimlen /work2/noaa/stmp/GFS_CI_ROOT/PR/1882/RUNTE\
STS/COMROT/C96_atm3DVar_a86b95ac/gdas.20211220/18//model_data/atmos/history/gdas.t18z.atmf006.nc pfull
Traceback (most recent call last):
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/gsi_utils.py", line 79, in get_ncdims
    import netCDF4 as nc
ModuleNotFoundError: No module named 'netCDF4'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/getncdimlen", line 16, in <module>
    FileDims = gsi_utils.get_ncdims(args.ncfile)
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/gsi_utils.py", line 81, in get_ncdims
    raise ImportError(f"Unable to import netCDF4 module\n{err}")
ImportError: Unable to import netCDF4 module
No module named 'netCDF4'
```

`getncdimlen` does not fail when executed under the rtreadon account in the 2021080100 gdasanal job (file `/work2/noaa/stmp/rtreadon/comrot/prgsinew/logs/2021080100/gdasanal.log`)
```
++ JGLOBAL_ATMOS_ANALYSIS[52]: /work2/noaa/da/rtreadon/git/global-workflow/intel2022_updates/ush/getncdimlen /work2/noaa/stmp/rtreadon/comrot/prgsinew/gdas.20210731/18//model_data/atmos/history/gdas.t18z.atmf006.nc pfull
+ JGLOBAL_ATMOS_ANALYSIS[52]: export LEVS=127
+ JGLOBAL_ATMOS_ANALYSIS[52]: LEVS=127
+ JGLOBAL_ATMOS_ANALYSIS[53]: status=0
```

I'm not sure where to go from here.  I can only test from my Orion account..  Orion tests under my account show that the changes in PR #1882 work.

WalterKolczynski-NOAA commented:
Orion has been having some lmod issues lately, to the point I opened a ticket with the helpdesk. Those have mostly been with `module avail` and `module spider` though.

Without looking at the log, this certainly sounds like the gfs_workflow python environment is not being loaded correctly, and maybe you are loading another python in your `.bashrc` that is providing NetCDF4. Will need to investigate.

WalterKolczynski-NOAA commented:
OH! Actually, that may be the issue, you aren't loading the gfs_workflow pyenv, are you, because you are using the modulefile in GSI. I'm not sure why it would work on your account and not under the CI.

WalterKolczynski-NOAA commented:
In any event, something is wrong with the python environment. Maybe @aerorahul has some ideas.

DavidHuber-NOAA commented:
I just confirmed that loading the GSI modulefile alone does not give access to the `netCDF4` python module needed for `getncdimlen`.  A kludge would be to activate the `regional_workflow` conda environment inside `anal.sh`, etc.  Otherwise, you may need to add the module use/load commands to access the miniconda installations on each machine specific to the global workflow (e.g. on Orion, `module use /work2/noaa/global/wkolczyn/save/hpc-stack/modulefiles/stack; module load hpc hpc-miniconda3/4.6.14 gfs_workflow/1.0.0`).

RussTreadon-NOAA commented:
Thanks @DavidHuber-NOAA and @WalterKolczynski-NOAA for identifying the problem.   I see the following in `jobs/rocoto/fcst.sh`
```
if [[ "${MACHINE_ID}" = "hera" ]]; then
  module use "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/modulefiles/core"
  module load "miniconda3/4.6.14"
  module load "gfs_workflow/1.0.0"
# TODO: orion and wcoss2 will be uncommented when they are ready.  This comment block will be removed in the next PR
#elif [[ "${MACHINE_ID}" = "orion" ]]; then
#  module use "/home/rmahajan/opt/global-workflow/modulefiles/core"
#  module load "python/3.7.5"
#  module load "gfs_workflow/1.0.0"
#elif [[ "${MACHINE_ID}" = "wcoss2" ]]; then
#  module load "python/3.7.5"
fi
```
Do you recommend adding an Orion block to `anal.sh`, `eobs.sh`, and `eupd.sh` akin to what is shown above for `fcst.sh`?

It is interesting that CI-Hera passed while CI-Orion failed.  Both ran the same `anal.sh`.   The Hera and Orion CI environments must differ.  

Should we update the Orion CI environment to mimic the Hera CI environment instead of adding more temporary scripting to `anal.sh`, `eobs.sh`, and `eupd.sh`?

WalterKolczynski-NOAA commented:
I think this is a difference in the base python installation on each machine.

Yes, mirror what we did for the forecast. The custom stack built by me is itself a kludge I'm hoping to finally get rid of soon when we get to spack-stack.

RussTreadon-NOAA commented:
Add the following to Orion `/work2/noaa/da/rtreadon/git/global-workflow/intel2022_updates/jobs/rocoto/eobs.sh`
```
@@ -30,6 +30,11 @@ else
     module use "${HOMEgfs}/sorc/gsi_enkf.fd/modulefiles"
     module load gsi_"${MACHINE_ID}"

+    if [[ "${MACHINE_ID}" = "orion" ]]; then
+        module use module use /work2/noaa/global/wkolczyn/save/hpc-stack/modulefiles/stack
+        module load hpc hpc-miniconda3/4.6.14 gfs_workflow/1.0.0
+    fi
+
     module list
     unset MACHINE_ID
     set_trace
```

Rewind and reboot enkfgdaseobs.   Job failed with

```
++ detect_machine.sh[68]: MACHINE_ID=orion
+ eobs.sh[15]: [[ orion = \w\c\o\s\s\2 ]]
+ eobs.sh[22]: COMPILER=intel
+ eobs.sh[23]: [[ orion = \h\e\r\a ]]
+ eobs.sh[23]: [[ orion = \c\h\e\y\e\n\n\e ]]
+ eobs.sh[28]: set +x
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.

Inactive Modules:
  1) hpc-intel/2022.1.2

Due to MODULEPATH changes, the following have been reloaded:
  1) hpc/1.2.0


The following have been reloaded with a version change:
  1) miniconda3/4.12.0 => miniconda3/4.6.14

/work2/noaa/global/wkolczyn/save/hpc-stack/core/miniconda3/4.6.14/etc/profile.d/conda.sh: line 55: PS1: unbound variable
End eobs.sh at 16:20:20 with error code 1 (time elapsed: 00:00:03)
```

Where do we go from here?

WalterKolczynski-NOAA commented:
Try the regional_workflow conda environment that Dave suggested. Using my stack isn't going to work because it unloads intel 2022, even if we fixed the PS1 issue.

RussTreadon-NOAA commented:
> Try the regional_workflow conda environment that Dave suggested. Using my stack isn't going to work because it unloads intel 2022, even if we fixed the PS1 issue.

I'm willing to try but I don't know what to add to `eobs.sh`.   As a start I added `conda list`.  `regional_workflow` was not in the list of available environments.  Sorry to be so dumb, but I'm not fluent in conda environments and modules.

RussTreadon-NOAA commented:
Might the following addition to `anal.sh`, `eobs.sh`, and `eupd.sh` work in the CI-Orion environment?

```
@@ -30,6 +30,14 @@ else
     module use "${HOMEgfs}/sorc/gsi_enkf.fd/modulefiles"
     module load gsi_"${MACHINE_ID}"

+    if [[ "${MACHINE_ID}" = "orion" ]]; then
+       wxflowPATH="${HOMEgfs}/ush/python:${HOMEgfs}/ush/python/wxflow/src"
+       PYTHONPATH="${PYTHONPATH:+${PYTHONPATH}:}${wxflowPATH}"
+       export PYTHONPATH
+    fi
+
     module list
     unset MACHINE_ID
     set_trace
```

RussTreadon-NOAA commented:
Any suggestions as to how to test various approaches without iteratively commiting and pushing changes to `RussTreadon-NOAA:feature/intel2022_updates` and asking the g-w team to trigger CI-Orion for each change?

RussTreadon-NOAA committed:
add workflow to PYTHONPATH on Orion (#1863)

DavidHuber-NOAA commented:
@RussTreadon-NOAA I am happy to test your changes so you don't have to commit/push to get CI to run.  This is what I think should work:

```
@@ -30,6 +30,14 @@ else
     module use "${HOMEgfs}/sorc/gsi_enkf.fd/modulefiles"
     module load gsi_"${MACHINE_ID}"

+    if [[ "${MACHINE_ID}" = "orion" ]]; then
+       conda activate regional_workflow
+    fi
+
     module list
     unset MACHINE_ID
     set_trace
```

RussTreadon-NOAA committed:
replace PYTHONPATH for Orion with regional_workflow (#1863)

RussTreadon-NOAA commented:
Thank you @DavidHuber-NOAA for the guidance.  The suggested change has been committed to `RussTreadon-NOAA:feature/intel2022_updates`.  

Done at [`d37d3d2`](https://github.com/NOAA-EMC/global-workflow/pull/1882/commits/d37d3d27cebd3ee3f5b1e1733ab57b3c48f953b9).

If you can trigger a CI-Orion test, that would be great!

RussTreadon-NOAA commented:
A check of `/work2/noaa/stmp/GFS_CI_ROOT/PR/1882/RUNTESTS/COMROT/C96_atm3DVar_d37d3d27/logs/2021122100/gdasanal.log` shows

```
++ detect_machine.sh[68]: MACHINE_ID=orion
+ anal.sh[15]: [[ orion = \w\c\o\s\s\2 ]]
+ anal.sh[22]: COMPILER=intel
+ anal.sh[23]: [[ orion = \h\e\r\a ]]
+ anal.sh[23]: [[ orion = \c\h\e\y\e\n\n\e ]]
+ anal.sh[28]: set +x
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
/apps/intel-2022.1.2/intel-2022.1.2/intelpython/latest/etc/conda/deactivate.d/xgboost_deactivate.sh: line 16: OCL_ICD_FILENAMES_RESET: unbound variable
End anal.sh at 23:16:01 with error code 1 (time elapsed: 00:00:09)
```

I did a rocotorewind and rocotoboot of my 2021080118 enkfgdas eobs.  It fails with the same error

```
++ conda.sh[56]: . /apps/intel-2022.1.2/intel-2022.1.2/intelpython/latest/etc/conda/deactivate.d/xgboost_deactivate.sh
/apps/intel-2022.1.2/intel-2022.1.2/intelpython/latest/etc/conda/deactivate.d/xgboost_deactivate.sh: line 16: OCL_ICD_FILENAMES_RESET: unbound variable
++++ xgboost_deactivate.sh[1]: postamble eobs.sh 1695774746 1
++++ preamble.sh[68]: set +x
End eobs.sh at 00:32:36 with error code 1 (time elapsed: 00:00:10)
```
   
Seems something else is needed to activate `regional_workflow`.

Any suggestions?

DavidHuber-NOAA commented:
@RussTreadon-NOAA I see now that we are missing a python/miniconda module load.  Orion is down today for maintenance, but when it comes back up tomorrow, I will check this out and get the solution to work.

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/intel2022_updates

DavidHuber-NOAA commented on jobs/rocoto/anal.sh:
```diff
@@ -4,9 +4,40 @@ source "${HOMEgfs}/ush/preamble.sh"
+    module use "${HOMEgfs}/sorc/gsi_enkf.fd/modulefiles"
+    module load gsi_"${MACHINE_ID}"
+
+    if [[ "${MACHINE_ID}" = "orion" ]]; then
+       conda activate regional_workflow
+    fi
```
Please try this and see if it works in CI:
```suggestion
    if [[ "${MACHINE_ID}" = "orion" ]]; then
       module load hpc-miniconda/4.12.0
       conda activate regional_workflow
    fi
```

RussTreadon-NOAA replied:
It is my understanding that developers are not permitted to trigger gw CI.   I did, however, make the suggested change and rewound and rebooted 2021080118 gdasanal from my Orion account.   The job failed with 
```
+ anal.sh[23]: [[ orion = \c\h\e\y\e\n\n\e ]]
+ anal.sh[28]: set +x
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
sh: line 1: 273469 Aborted                 /bin/lua /apps/lmod/lmod/libexec/lmod bash --regexp --no_redirect --spider_timeout 2.0 spider '^hpc%-miniconda/4%.12%.0$' 2> /dev/null
sh: line 1: 273472 Aborted                 /bin/lua /apps/lmod/lmod/libexec/lmod bash --regexp --no_redirect --spider_timeout 2.0 spider '^hpc%-miniconda/4%.12%.0$' 2> /dev/null
Lmod has detected the following error: These module(s) or extension(s) exist
but cannot be loaded as requested: "hpc-miniconda/4.12.0"
   Try: "module spider hpc-miniconda/4.12.0" to see how to load the module(s).
```
From previous runs in which I did not attempt to activate a conda environment, the following modules are loaded by `gsi_orion.lua` and `gsi_common.lua`
```
Currently Loaded Modules:
  1) contrib/0.1         10) netcdf/4.7.4  19) wrf_io/1.2.0
  2) miniconda3/4.12.0   11) bufr/11.7.0   20) ncio/1.1.2
  3) hpc/1.2.0           12) bacio/2.4.1   21) crtm/2.4.0
  4) intel/2022.1.2      13) w3emc/2.9.2   22) ncdiag/1.1.1
  5) hpc-intel/2022.1.2  14) sp/2.3.3      23) gsi_common
  6) impi/2022.1.2       15) ip/3.3.3      24) prod_util/1.2.2
  7) hpc-impi/2022.1.2   16) sigio/2.3.2   25) gsi_orion
  8) cmake/3.22.1        17) sfcio/1.4.1
  9) hdf5/1.10.6         18) nemsio/2.5.4
```

DavidHuber-NOAA replied:
@RussTreadon-NOAA My apologies and thanks for the clarification on CI.  I had a typo in the line I added.  It should be

`module load hpc-miniconda3/4.12.0`

That said, I'm starting to think this won't work, either.  Feel free to give it a try anyway, but I will also run some tests.

RussTreadon-NOAA replied:
As you suspect, adding "3" did not yield a successful run.  I'm back to the same `unbound variable` error as before

```
/apps/intel-2022.1.2/intel-2022.1.2/intelpython/latest/etc/conda/deactivate.d/xgboost_deactivate.sh: line 16: OCL_ICD_FILENAMES_RESET: unbound variable
End anal.sh at 14:02:23 with error code 1 (time elapsed: 00:00:08)
```

Since CI-Orion originally died in `gdasanal` with `getncdimlen` saying that it could not find `gsi_utils`, what if we update `PYTHONPATH` to include the path to `gsi_utils`.   
```
     if [[ "${MACHINE_ID}" = "orion" ]]; then
-       conda activate regional_workflow
+      wxflowPATH="${HOMEgfs}/ush:${HOMEgfs}/ush/python:${HOMEgfs}/ush/python/wxflow/src"
+      PYTHONPATH="${PYTHONPATH:+${PYTHONPATH}:}${wxflowPATH}"
+      export PYTHONPATH
     fi
```
`gsi_utils.py` resides in `$HOMEgfs/ush`.

The above change to `anal.sh` works when I rocotoboot 2021080118 gdasanal from my Orion account.  I don't know if this change would work from the account which runs CI-Orion.

DavidHuber-NOAA commented:
@RussTreadon-NOAA Looking at the CI error, it seems the issue was that it could not load the netCDF4 module required by `getncdimlen`:

```
++ JGLOBAL_ATMOS_ANALYSIS[52]: /work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/getncdimlen /work2/noaa/stmp/GFS_CI_ROOT/PR/1882/RUNTE\
STS/COMROT/C96_atm3DVar_a86b95ac/gdas.20211220/18//model_data/atmos/history/gdas.t18z.atmf006.nc pfull
Traceback (most recent call last):
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/gsi_utils.py", line 79, in get_ncdims
    import netCDF4 as nc
ModuleNotFoundError: No module named 'netCDF4'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/getncdimlen", line 16, in <module>
    FileDims = gsi_utils.get_ncdims(args.ncfile)
  File "/work2/noaa/stmp/GFS_CI_ROOT/PR/1882/global-workflow/ush/gsi_utils.py", line 81, in get_ncdims
    raise ImportError(f"Unable to import netCDF4 module\n{err}")
ImportError: Unable to import netCDF4 module
No module named 'netCDF4'
```
I tried adding your suggestion (i.e. adding ush to the PYTHONPATH), but I am still getting the netCDF4 error above.

RussTreadon-NOAA commented:
OK.  Thank you for trying and clearly identifying the problem.  This is odd.  CI-Hera works.  CI-Orion fails.   

A diff of `gsi_hera.intel.lua` and `gsi_orion.lua` in `sorc/gsi_enkf.fd/modulefiles` returns
```
Orion-login-3:/work2/noaa/da/rtreadon/git/global-workflow/intel2022_updates/sorc/gsi_enkf.fd/modulefiles$ diff gsi_hera.intel.lua gsi_orion.lua
4c4
< prepend_path("MODULEPATH", "/scratch1/NCEPDEV/nems/role.epic/miniconda3/modulefiles")
---
> prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/contrib/orion/miniconda3/modulefiles")
8c8
< prepend_path("MODULEPATH", "/scratch1/NCEPDEV/nems/role.epic/hpc-stack/libs/intel-2022.1.2/modulefiles/stack")
---
> prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/contrib/orion/hpc-stack/intel-2022.1.2/modulefiles/stack")
13c13
< local cmake_ver=os.getenv("cmake_ver") or "3.20.1"
---
> local cmake_ver=os.getenv("cmake_ver") or "3.22.1"
28c28
< pushenv("GSI_BINARY_SOURCE_DIR", "/scratch1/NCEPDEV/global/glopara/fix/gsi/20230911")
---
> pushenv("GSI_BINARY_SOURCE_DIR", "/work/noaa/global/glopara/fix/gsi/20230911")
30c30
< whatis("Description: GSI environment on Hera with Intel Compilers")
---
> whatis("Description: GSI environment on Orion with Intel Compilers")
```

Are the `miniconda3/modulefiles` and `hpc-stack/intel-2022.1.2` consistent between Hera and Orion?

Is the environment from which CI-Hera runs consistent with the environment from which CI-Orion runs?

This PR is stuck due to CI-Orion.  I don't know what to change to get this PR unstuck.

DavidHuber-NOAA commented:
@RussTreadon-NOAA I don't think that they are consistent between the two machines.  It seems odd to me that when attempting to activate regional_workflow that the intelpython `conda deactivate` is called (which is where the `OCL_ICD_FILENAMES_RESET: unboud variable` error is copping up at.  This may be a system bug.

Indeed, when on a terminal and I load gsi_orion.lua, several python variables are set to intelpython locations.  In particular: `CONDA_PREFIX=/apps/intel-2022.1.2/intel-2022.1.2/intelpython/latest`.  I'm working on tracking down where this comes from.

DavidHuber-NOAA commented:
The conda environmental variables pointing to intelpython are set when the intel/2022.1.2 modulefile is loaded.  I believe this is the source of the errors we are seeing when attempting to activate the regional_workflow environment.  I will open an issue with RDHPCS on this.

Unfortunately, that does not solve our current issue.  I recall now that the regional workflow loads this conda environment at runtime and on investigating see that they wrap it in `set +u ... set -u`, probably to get around this issue.  Lastly, it appears the regional_workflow environment doesn't have netCDF4, either.  Instead I had to load the regional_workflow_cmaq environment.  After all of this, the gdasanal job now runs for me.  So, give this a try and I will kick off the CI jobs:

```
    if [[ "${MACHINE_ID}" = "orion" ]]; then
       module load miniconda3/4.12.0
       set +u
       conda activate regional_workflow_cmaq
       set -u
    fi
```

RussTreadon-NOAA committed:
replace regional_workflow with regional_workflow_cmaq (#1863)

RussTreadon-NOAA commented:
Replace `regional_workflow` with `regional_workflow_cmaq`.   Add `set +u` and `set -u` around `conda activate`.  Done at 95b705f.

RussTreadon-NOAA commented:
Thank you @DavidHuber-NOAA for troubleshooting the CI-Orion problem and suggesting a fix.   CI on Orion and Hera just passed!

@WalterKolczynski-NOAA , are there any other changes you need from me for this PR?

WalterKolczynski-NOAA commented on jobs/rocoto/anal.sh:
```diff
@@ -4,9 +4,43 @@ source "${HOMEgfs}/ush/preamble.sh"
+       module load miniconda3/4.12.0
+       set +u
+       conda activate regional_workflow_cmaq
+       set -u
```
We have a function setup by the preamble to handle turning this back on. This way it isn't turned on if `$STRICT` was `"NO"`.
```suggestion
       set_strict
```

WalterKolczynski-NOAA commented on jobs/rocoto/anal.sh:
```diff
@@ -4,9 +4,43 @@ source "${HOMEgfs}/ush/preamble.sh"
+       module load miniconda3/4.12.0
+       set +u
+       conda activate regional_workflow_cmaq
+       set -u
```
Similar to `set_trace`, we have a function setup by the preamble to handle turning this back on. This way it isn't turned on if `$STRICT` was `"NO"`.
```suggestion
       set_strict
```

RussTreadon-NOAA replied:
`set -u` replaced with `set_strict` at 33846f5

WalterKolczynski-NOAA commented on jobs/rocoto/eobs.sh:
```diff
@@ -4,9 +4,43 @@ source "${HOMEgfs}/ush/preamble.sh"
+       module load miniconda3/4.12.0
+       set +u
+       conda activate regional_workflow_cmaq
+       set -u
```
```suggestion
       set_strict
```

RussTreadon-NOAA replied:
`set -u` replaced with `set_strict` at 33846f5

WalterKolczynski-NOAA commented on jobs/rocoto/eupd.sh:
```diff
@@ -4,9 +4,43 @@ source "${HOMEgfs}/ush/preamble.sh"
+       module load miniconda3/4.12.0
+       set +u
+       conda activate regional_workflow_cmaq
+       set -u
```
```suggestion
       set_strict
```

RussTreadon-NOAA replied:
`set -u` replaced with `set_strict` at 33846f5

RussTreadon-NOAA committed:
replace set -u with set_strict (#1863)

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/intel2022_updates

RussTreadon-NOAA commented:
`RussTreadon-NOAA:feature/intel2022_updates` brought up to date with the current head (3e750d1) of `develop`.  Done at b00f0a7.   

2021080118 gdasanal, enkfgdaseobs, and enkfgdaseupd rewound and rebooted on Orion.  Each job successfully ran to completion.

Would EID staff, @DavidHuber-NOAA or @WalterKolczynski-NOAA , please rerun CI tests on Hera and Orion
